src/Nethermind/Nethermind.Merge.Plugin/Handlers/ExecutePayloadV1Handler.cs

```diff
@@ -0,0 +1,206 @@
+using Nethermind.JsonRpc;
+using Nethermind.Logging;
+using Nethermind.Merge.Plugin.Data;
+
+namespace Nethermind.Merge.Plugin.Handlers
+{
+    /// <summary>
+    /// https://hackmd.io/@n0ble/kintsugi-spec
+    /// Verifies the payload according to the execution environment rule set (EIP-3675)
+    /// and returns the status of the verification and the hash of the last valid block
+    /// </summary>
+    public class ExecutePayloadV1Handler : IHandler<BlockRequestResult, ExecutePayloadV1Result>
+    {
+        private const string AndWontBeAcceptedToTheTree = "and wont be accepted to the tree";
+        private readonly IHeaderValidator _headerValidator;
+        private readonly IBlockTree _blockTree;
+        private readonly IBlockchainProcessor _processor;
+        private readonly IEthSyncingInfo _ethSyncingInfo;
+        private readonly IInitConfig _initConfig;
+        private readonly ILogger _logger;
+        private SemaphoreSlim _blockValidationSemaphore;
+        private readonly LruCache<Keccak, bool> _latestBlocks = new(50, "LatestBlocks");
+
+        public ExecutePayloadV1Handler(
+            IHeaderValidator headerValidator,
+            IBlockTree blockTree,
+            IBlockchainProcessor processor,
+            IEthSyncingInfo ethSyncingInfo,
+            IInitConfig initConfig,
+            ILogManager logManager)
+        {
+            _headerValidator = headerValidator ?? throw new ArgumentNullException(nameof(headerValidator));
+            _blockTree = blockTree;
+            _processor = processor;
+            _ethSyncingInfo = ethSyncingInfo;
+            _initConfig = initConfig;
+            _logger = logManager.GetClassLogger();
+            _blockValidationSemaphore = new SemaphoreSlim(0);
+            _processor.BlockProcessed += (s, e) =>
+            {
+                _blockValidationSemaphore.Release(1);
+            };
+            _processor.BlockInvalid += (s, e) =>
+            {
+                _blockValidationSemaphore.Release(1);
+            };
+        }
+
+        public ResultWrapper<ExecutePayloadV1Result> Handle(BlockRequestResult request)
+        {
+            ExecutePayloadV1Result executePayloadResult = new();
+
+            if (_ethSyncingInfo.IsSyncing())
+            {
+                executePayloadResult.EnumStatus = VerificationStatus.Syncing;
+                return ResultWrapper<ExecutePayloadV1Result>.Success(executePayloadResult);
+            }
+
+            ValidationResult result = ValidateRequestAndProcess(request, out Block? processedBlock);
+            if ((result & ValidationResult.AlreadyKnown) != 0 || result == ValidationResult.Invalid)
+            {
+                bool isValid = (result & ValidationResult.Valid) != 0;
+                executePayloadResult.EnumStatus = isValid ? VerificationStatus.Valid : VerificationStatus.Invalid;
+                executePayloadResult.LatestValidHash = isValid ? request.BlockHash : _blockTree.HeadHash;
```

Comment:
not sure if `_blockTree.HeadHash` retrieves the hash of the last valid block

Reply:
Yeah, good catch. I am not sure too. Let's ask specification authors.